swconfig/files/switch.lua

```diff
@@ -0,0 +1,438 @@
+    if char ~= char_echo and (self.state ~= Switch.State.LOGIN_PASSWORD or char_echo ~= '*') then
+      print(string.format("Command echo error: got '%s', expected '%s'", char_echo, char))
+      self.sock:flush()
+    end
+  end
+
+  return true
+end
+
+-- Implement serial read without the need to specify a length
+-- Read as much bytes as there is
+-- We need to do this because the underlying serial lib doesn't handle EOF
+-- EOF should be possible because the switch uses one stop bit in its UART implementation
+-- But because we don't have any EOF mechanism, we assume there's nothing to read if we hit a timeout error
+-- A timeout can be specified. If not specified, default timeout apply.
+-- The timeout is only used when waiting for the first char of the response.
+-- Once the first char arrived, we become more strict and the timeout becomes the default again
+function Switch:_recv(timeout)
+  res = ""
+
+  -- When user doesn't specify any timeout, use default value
+  if not timeout then timeout = SERIAL_READ_ONE_CHAR_TIMEOUT end
+
+  while true do
+    -- Reset the timeout to the default one once the first char has been obtained
+    if res:len() == 1 then
+      timeout = SERIAL_READ_ONE_CHAR_TIMEOUT
+    end
+
+    err, data, len = self.sock:read(1, timeout, 0)
+
+    if err ~= rs232.RS232_ERR_NOERROR or len == 0 then
+      break
+    end
+    res = res .. data
+  end
+
+  if res:len() == 0 then
+    self:_print_error("Error: No data has been received", err)
+    return nil
+  end
+
+  -- TODO: Handle the --More--
+
+  return res
+end
+
+-- This method analyzes the received output to determine the switch state
+function Switch:_parse_prompt(data)
+  first_line = data[1]
+  last_line = data[#data]
+
+  -- It is crucial below to use "starts" and not "ends" because sometimes, there will be a comment after the prompt
+  -- For example:
+  --  martinsw# *Jan 08 2000 03:54:00: %System-5: New console connection for user admin, source async  ACCEPTED
+  --  There should never be anything BEFORE the prompt
+  if string.starts(last_line, "Username: ") then
+    print("State Username detected")
+    return Switch.State.LOGIN_USERNAME
+
+  elseif string.starts(last_line, "Password: ") then
+    print("State Password detected")
+    return Switch.State.LOGIN_PASSWORD
+
+    -- Press any key to continue is not the last line but the first one
+    --  Press any key to continue
+    --  *Jan 08 2000 06:42:48: %System-5: New console connection for user admin, source async  REJECTED
+  elseif first_line == "Press any key to continue" then
+    print("State 'Press any key' detected")
+    return Switch.State.PRESS_ANY_KEY
+  end
+
+  -- Determine the hostname
+  if self.hostname == nil then
+    -- '%' is the lua escape character for pattern, because '(' otherwise would have another meaning
+    needles = {"%(", "# ", "> "}
+    for key, needle in ipairs(needles) do
+      hostname_end = string.find(last_line, needle)
```

Comment:
`first_char, _  := string.find(.....)` to be clear that you ignore the last return